quarkc/ruby.py

```diff
@@ -0,0 +1,412 @@
+    assert False
+
+def main(statements):
+    return '\nif __FILE__ == $0%send\n\n' % block(statements)
+
+## Naming and imports
+
+SUBS = {'Class': 'QuarkClass', 'end': 'end_', 'next': 'next_'}
+def name(n):
+    return SUBS.get(n, n).replace('-', '_')
+
+def type(path, name, parameters):
+    return ".".join(path + [name])
+
+def import_(path, origin, dep, cache={}):
+    if dep is None:
+        # common 'directories'
+        common = len(tuple(itertools.takewhile(
+            lambda p: p[0]==p[1], zip(path[:-1], origin[:-1]))))
+        # uncommon path directories
+        lpath = path[common:-1]
+        # uncommon origin directories
+        lorigin = origin[common:-1]
+        # go up and down and load the file
+        rpath = "/".join(('..',) * len(lorigin) + lpath + path[-1:])
+        require = "require_relative '%s' # %s %s %s" % (rpath, common, lpath, lorigin)
+        if origin == ('quark_md',) and path != ('quark', 'reflect', ):
+            # XXX: why does quark think that quark_md depends on quark.concurrent and quark.behavior ???
+            require = "# for quark_md: %s" % (require, )
+        elif len(origin) == 1 and origin[0].endswith("_md") and origin[0] != 'quark_md':
+            require = "# for %s: %s" % (origin, require, )
+        return require
+    else:
+        if len(origin) == 1:
+            if len(path) == 1:
+                return "require '%s' " % path[0]
+            else:
+                return "require '%s' # .../%s" % (path[0], "/".join(path[1:]))
+        else:
+            return "require '%s' # .../%s %s" % (path[0], "/".join(path[1:]), "/".join(origin))
+
+def qualify(package, origin):
+    # Always fully-qualify names, because Ruby is not fully lexically-scoped.
+    return package
+
+## Documentation
+
+def doc(lines):
+    return doc_helper(lines, '#  ', '#  ', '#  ')
+
+## Comments
+
+def comment(stuff):
+    return '# %s\n' % stuff
+
+## Class definition
+
+def clazz(doc, abstract, name, parameters, base, interfaces, static_fields, fields, constructors, methods):
+    rname = 'CLASS_' + name
+    prologue = []
+    if fields:
+        prologue.extend(['attr_accessor %s' % ', '.join(':' + name for name, value in fields)])
+    postscript = []
+    if static_fields:
+        prologue.extend(['extend ::DatawireQuarkCore::Static', ''])
+        prologue.extend(static_fields)
+        postscript.append('%s.unlazy_statics' % rname)
+        
+    init_fields = Templates.method(
+        name='__init_fields__',
+        parameters='',
+        body=indent(''.join('\nself.%s = %s' % pairs for pairs in fields)),
+    )
+    source = Templates.class_(
+        name=rname,
+        alias=name,
+        base=('::Quark.' + base) if base else '::DatawireQuarkCore::QuarkObject',
+        prologue=indent(prologue, leading_nl=False),
+        constructors=indent(constructors),
+        methods=indent(methods + [init_fields]),
+        postscript=indent(postscript, level=0, leading_nl=False),
```

Comment:
indent learned the `"\n".join(arg) on the way. and `indent=0` lines it up with the outer body of the class